cirq/experiments/fidelity_estimation.py

```diff
@@ -372,3 +394,225 @@ def least_squares_xeb_fidelity_from_probabilities(
+        records = []
+        for task, n_result in zip(tasks, results):
+            assert len(n_result) == 1, len(n_result)
+            result = n_result[0]  # sweep nesting
+            sampled_inds = result.data.values[:, 0]
+            sampled_probs = np.bincount(sampled_inds, minlength=2 ** 2) / len(sampled_inds)
+
+            records += [
+                {
+                    'circuit_i': task.circuit_i,
+                    'cycle_depth': task.cycle_depth,
+                    'sampled_probs': sampled_probs,
+                }
+            ]
+        return records
+
+
+def sample_2q_xeb_circuits(
+    sampler: 'cirq.Sampler',
+    circuits: Sequence['cirq.Circuit'],
+    cycle_depths: Sequence[int],
+    repetitions: int = 10_000,
+    batch_size: int = 9,
+):
+    """Sample two-qubit XEB circuits given a sampler.
+
+    Args:
+        sampler: A Cirq sampler for executing circuits.
+        circuits: A library of two-qubit circuits generated from
+            `random_rotations_between_two_qubit_circuit` of sufficient length
+            for `cycle_depths`.
+        cycle_depths: A sequence of cylce depths at which we will truncate each
+            of the `circuits` to execute.
+        repetitions: Each (circuit, cycle_depth) will be sampled for this many
+            repetitions.
+        batch_size: We call `run_batch` on the sampler, which can speed up
+            execution in certain environments. The number of
+            (circuit, cycle_depth) tasks to be run in each batch
+            is given by this number.
+
+    Returns:
+        A pandas dataframe with index given by ['circuit_i', 'cycle_depth'] and
+        column "sampled_probs".
+    """
+    import tqdm
+
+    all_all_qubits_s: Set['cirq.Qid'] = set()
+    all_all_qubits_s = all_all_qubits_s.union(*(circuit.all_qubits() for circuit in circuits))
+    all_all_qubits = sorted(all_all_qubits_s)
+    if len(all_all_qubits) != 2:
+        raise ValueError(
+            "`circuits` should be a sequence of circuits each operating on the same two qubits."
+        )
+    q0, q1 = all_all_qubits
+
+    tasks = []
+    for cycle_depth in cycle_depths:
+        for circuit_i, circuit in enumerate(circuits):
+            circuit_depth = cycle_depth * 2 + 1
+            assert circuit_depth <= len(circuit)
+            tcircuit = circuit[:circuit_depth]
+            pcircuit = tcircuit + ops.measure(q0, q1)
+            tasks.append(
+                _Sample2qXEBTask(
+                    cycle_depth=cycle_depth, circuit_i=circuit_i, prepared_circuit=pcircuit
+                )
+            )
+
+    n_tasks = len(tasks)
+    b_tasks = [tasks[i : i + batch_size] for i in range(0, n_tasks, batch_size)]
+
+    run_batch = _SampleInBatches(sampler=sampler, repetitions=repetitions)
+    with ThreadPoolExecutor(max_workers=2) as pool:
+        futures = [pool.submit(run_batch, task) for task in b_tasks]
+
+        records = []
+        with tqdm.tqdm(total=n_tasks) as progress:
```

Comment:
Have you checked that this work in all circuimstances, when this is executed as a script or test?

I always assumed the pattern of selecting the progress framework to the end user of a library. This allows this functionality to be composed in a better way. What might be a good compromise to have a "report_progress" argument for this function.

Reply:
First, yes it works as a script, test, or notebook. 

But! Point taken. Yes, usually there should be some sort of optional progress callback. I think for this function in particular it would be nice to have it to default to having a nice progress bar. This is one of the issues with having a function that rides the line between "library" code and "application" code all in Cirq. Curious what @balopat or @maffoo would think about putting a progress bar here

Reply:
pushed a commit with more-optional tqdm, but tqdm by default